### PR TITLE
fix(runtime): autosave engine-implicit RuntimeCache via atexit + weakref

### LINKED ISSUE #4359: 🐛 [Bug] Serialization fails in Runtime Cache
Running the following script on commit `a094c74395cd2ab50621c866f25a05f761b4931c`:

```
from pathlib import Path

import torch
import torch_tensorrt as torchtrt
import torchvision.models as models

torch.manual_seed(0)
torch.cuda.manual_seed_all(0)

artifact = Path(__file__).resolve().parent / "trt_mod.pt2"


def main():
    model = models.resnet50(pretrained=False).eval().to("cuda")

    inputs = torch.randn(1, 3, 224, 224, device="cuda")

    exported = torch.export.export(model, (inputs,))

    trt_mod = torchtrt.dynamo.compile(
        exported,
        inputs=[inputs],
        min_block_size=1,
        immutable_weights=False,
        reuse_cached_engines=False,
        enabled_precisions={torch.float32},
        torch_executed_ops={torch.ops.aten.relu.default},
    )

    with torch.no_grad():
        results1 = trt_mod(inputs)

    torchtrt.save(
        trt_mod,
        str(artifact),
        output_format="exported_program",
        inputs=[inputs],
        retrace=False,
    )

    trt_ep = torch.export.load(str(artifact))

    # Inspect loaded engine types
    gm = trt_ep.module()

    with torch.no_grad():
        results2 = gm(inputs)

    print("max diff:", (results1 - results2).abs().max().item())
    # for name, val in gm._modules.items():
    #     print(f"  submodule {name}: {type(val).__name__}")
    # for attr_name in dir(gm):
    #     obj = getattr(gm, attr_name, None)
    #     if obj is not None and "engine" in attr_name.lower():
    #         print(f"  attr {attr_name}: {type(obj)} is_script_obj={isinstance(obj, torch._C.ScriptObject)}")
    # for fqn, const in trt_ep.constants.items():
    #     print(f"  constant {fqn}: {type(const)} is_script_obj={isinstance(const, torch._C.ScriptObject)}")
    # for node in gm.graph.nodes:
    #     if node.op == "get_attr":
    #         val = getattr(gm, node.target, None)
    #         print(f"  graph get_attr '{node.target}': {type(val)}")


if __name__ == "__main__":
    main()


```

Gives the error:

```
WARNING: [Torch-TensorRT - Debug Build] - RuntimeCacheHandle::serialize() invoked on a non-RTX build; returning empty bytes.
WARNING: [Torch-TensorRT - Debug Build] - RuntimeCacheHandle::serialize() invoked on a non-RTX build; returning empty bytes.
WARNING: [Torch-TensorRT - Debug Build] - RuntimeCacheHandle::serialize() invoked on a non-RTX build; returning empty bytes.
WARNING: [Torch-TensorRT - Debug Build] - RuntimeCacheHandle::serialize() invoked on a non-RTX build; returning empty bytes.
Traceback (most recent call last):
  File "/home/other/test_two_runtime.py", line 65, in <module>
    main()
  File "/home/other/test_two_runtime.py", line 33, in main
    torchtrt.save(
  File "/home/TensorRT/py/torch_tensorrt/_compile.py", line 1001, in save
    exp_program = export(
                  ^^^^^^^
  File "/home/TensorRT/py/torch_tensorrt/dynamo/_exporter.py", line 42, in export
    patched_module = transform(gm, cross_compile_module)
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/TensorRT/py/torch_tensorrt/dynamo/_exporter.py", line 92, in transform
    gm.graph.eliminate_dead_code()
  File "/usr/local/lib/python3.12/dist-packages/torch/fx/graph.py", line 2652, in eliminate_dead_code
    self.lint()
  File "/usr/local/lib/python3.12/dist-packages/torch/fx/graph.py", line 2559, in lint
    check_arg(arg, node)
  File "/usr/local/lib/python3.12/dist-packages/torch/fx/graph.py", line 2544, in check_arg
    raise RuntimeError(
RuntimeError: Argument 'getitem' of Node 'relu' was used before it has been defined! Please check that Nodes in the graph are topologically ordered
...
    return (getitem_53,)
Exception ignored in: <function RuntimeCache.__del__ at 0x7fe862091b20>
Traceback (most recent call last):
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 309, in __del__
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 204, in path
ImportError: sys.meta_path is None, Python is likely shutting down
Exception ignored in: <function RuntimeCache.__del__ at 0x7fe862091b20>
Traceback (most recent call last):
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 309, in __del__
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 204, in path
ImportError: sys.meta_path is None, Python is likely shutting down
Exception ignored in: <function RuntimeCache.__del__ at 0x7fe862091b20>
Traceback (most recent call last):
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 309, in __del__
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 204, in path
ImportError: sys.meta_path is None, Python is likely shutting down
Exception ignored in: <function RuntimeCache.__del__ at 0x7fe862091b20>
Traceback (most recent call last):
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 309, in __del__
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 204, in path
ImportError: sys.meta_path is None, Python is likely shutting down
Exception ignored in: <function RuntimeCache.__del__ at 0x7fe862091b20>
Traceback (most recent call last):
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 309, in __del__
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 204, in path
ImportError: sys.meta_path is None, Python is likely shutting down
Exception ignored in: <function RuntimeCache.__del__ at 0x7fe862091b20>
Traceback (most recent call last):
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 309, in __del__
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 204, in path
ImportError: sys.meta_path is None, Python is likely shutting down
Exception ignored in: <function RuntimeCache.__del__ at 0x7fe862091b20>
Traceback (most recent call last):
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 309, in __del__
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 204, in path
ImportError: sys.meta_path is None, Python is likely shutting down
Exception ignored in: <function RuntimeCache.__del__ at 0x7fe862091b20>
Traceback (most recent call last):
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 309, in __del__
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 204, in path
ImportError: sys.meta_path is None, Python is likely shutting down
Exception ignored in: <function RuntimeCache.__del__ at 0x7fe862091b20>
Traceback (most recent call last):
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 309, in __del__
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 204, in path
ImportError: sys.meta_path is None, Python is likely shutting down
Exception ignored in: <function RuntimeCache.__del__ at 0x7fe862091b20>
Traceback (most recent call last):
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 309, in __del__
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 204, in path
ImportError: sys.meta_path is None, Python is likely shutting down
Exception ignored in: <function RuntimeCache.__del__ at 0x7fe862091b20>
Traceback (most recent call last):
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 309, in __del__
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 204, in path
ImportError: sys.meta_path is None, Python is likely shutting down
Exception ignored in: <function RuntimeCache.__del__ at 0x7fe862091b20>
Traceback (most recent call last):
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 309, in __del__
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 204, in path
ImportError: sys.meta_path is None, Python is likely shutting down
Exception ignored in: <function RuntimeCache.__del__ at 0x7fe862091b20>
Traceback (most recent call last):
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 309, in __del__
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 204, in path
ImportError: sys.meta_path is None, Python is likely shutting down
Exception ignored in: <function RuntimeCache.__del__ at 0x7fe862091b20>
Traceback (most recent call last):
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 309, in __del__
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 204, in path
ImportError: sys.meta_path is None, Python is likely shutting down
Exception ignored in: <function RuntimeCache.__del__ at 0x7fe862091b20>
Traceback (most recent call last):
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 309, in __del__
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 204, in path
ImportError: sys.meta_path is None, Python is likely shutting down
Exception ignored in: <function RuntimeCache.__del__ at 0x7fe862091b20>
Traceback (most recent call last):
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 309, in __del__
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 204, in path
ImportError: sys.meta_path is None, Python is likely shutting down
Exception ignored in: <function RuntimeCache.__del__ at 0x7fe862091b20>
Traceback (most recent call last):
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 309, in __del__
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 204, in path
ImportError: sys.meta_path is None, Python is likely shutting down
Exception ignored in: <function RuntimeCache.__del__ at 0x7fe862091b20>
Traceback (most recent call last):
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 309, in __del__
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 204, in path
ImportError: sys.meta_path is None, Python is likely shutting down
Exception ignored in: <function RuntimeCache.__del__ at 0x7fe862091b20>
Traceback (most recent call last):
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 309, in __del__
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 204, in path
ImportError: sys.meta_path is None, Python is likely shutting down
Exception ignored in: <function RuntimeCache.__del__ at 0x7fe862091b20>
Traceback (most recent call last):
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 309, in __del__
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 204, in path
ImportError: sys.meta_path is None, Python is likely shutting down
Exception ignored in: <function RuntimeCache.__del__ at 0x7fe862091b20>
Traceback (most recent call last):
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 309, in __del__
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 204, in path
ImportError: sys.meta_path is None, Python is likely shutting down
Exception ignored in: <function RuntimeCache.__del__ at 0x7fe862091b20>
Traceback (most recent call last):
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 309, in __del__
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 204, in path
ImportError: sys.meta_path is None, Python is likely shutting down
Exception ignored in: <function RuntimeCache.__del__ at 0x7fe862091b20>
Traceback (most recent call last):
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 309, in __del__
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 204, in path
ImportError: sys.meta_path is None, Python is likely shutting down
Exception ignored in: <function RuntimeCache.__del__ at 0x7fe862091b20>
Traceback (most recent call last):
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 309, in __del__
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 204, in path
ImportError: sys.meta_path is None, Python is likely shutting down
Exception ignored in: <function RuntimeCache.__del__ at 0x7fe862091b20>
Traceback (most recent call last):
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 309, in __del__
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 204, in path
ImportError: sys.meta_path is None, Python is likely shutting down
Exception ignored in: <function RuntimeCache.__del__ at 0x7fe862091b20>
Traceback (most recent call last):
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 309, in __del__
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 204, in path
ImportError: sys.meta_path is None, Python is likely shutting down
Exception ignored in: <function RuntimeCache.__del__ at 0x7fe862091b20>
Traceback (most recent call last):
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 309, in __del__
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 204, in path
ImportError: sys.meta_path is None, Python is likely shutting down
Exception ignored in: <function RuntimeCache.__del__ at 0x7fe862091b20>
Traceback (most recent call last):
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 309, in __del__
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 204, in path
ImportError: sys.meta_path is None, Python is likely shutting down
Exception ignored in: <function RuntimeCache.__del__ at 0x7fe862091b20>
Traceback (most recent call last):
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 309, in __del__
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 204, in path
ImportError: sys.meta_path is None, Python is likely shutting down
Exception ignored in: <function RuntimeCache.__del__ at 0x7fe862091b20>
Traceback (most recent call last):
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 309, in __del__
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 204, in path
ImportError: sys.meta_path is None, Python is likely shutting down
Exception ignored in: <function RuntimeCache.__del__ at 0x7fe862091b20>
Traceback (most recent call last):
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 309, in __del__
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 204, in path
ImportError: sys.meta_path is None, Python is likely shutting down
Exception ignored in: <function RuntimeCache.__del__ at 0x7fe862091b20>
Traceback (most recent call last):
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 309, in __del__
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 204, in path
ImportError: sys.meta_path is None, Python is likely shutting down
Exception ignored in: <function RuntimeCache.__del__ at 0x7fe862091b20>
Traceback (most recent call last):
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 309, in __del__
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 204, in path
ImportError: sys.meta_path is None, Python is likely shutting down
Exception ignored in: <function RuntimeCache.__del__ at 0x7fe862091b20>
Traceback (most recent call last):
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 309, in __del__
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 204, in path
ImportError: sys.meta_path is None, Python is likely shutting down
Exception ignored in: <function RuntimeCache.__del__ at 0x7fe862091b20>
Traceback (most recent call last):
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 309, in __del__
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 204, in path
ImportError: sys.meta_path is None, Python is likely shutting down
Exception ignored in: <function RuntimeCache.__del__ at 0x7fe862091b20>
Traceback (most recent call last):
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 309, in __del__
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 204, in path
ImportError: sys.meta_path is None, Python is likely shutting down
Exception ignored in: <function RuntimeCache.__del__ at 0x7fe862091b20>
Traceback (most recent call last):
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 309, in __del__
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 204, in path
ImportError: sys.meta_path is None, Python is likely shutting down
Exception ignored in: <function RuntimeCache.__del__ at 0x7fe862091b20>
Traceback (most recent call last):
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 309, in __del__
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 204, in path
ImportError: sys.meta_path is None, Python is likely shutting down
Exception ignored in: <function RuntimeCache.__del__ at 0x7fe862091b20>
Traceback (most recent call last):
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 309, in __del__
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 204, in path
ImportError: sys.meta_path is None, Python is likely shutting down
Exception ignored in: <function RuntimeCache.__del__ at 0x7fe862091b20>
Traceback (most recent call last):
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 309, in __del__
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 204, in path
ImportError: sys.meta_path is None, Python is likely shutting down
Exception ignored in: <function RuntimeCache.__del__ at 0x7fe862091b20>
Traceback (most recent call last):
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 309, in __del__
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 204, in path
ImportError: sys.meta_path is None, Python is likely shutting down
Exception ignored in: <function RuntimeCache.__del__ at 0x7fe862091b20>
Traceback (most recent call last):
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 309, in __del__
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 204, in path
ImportError: sys.meta_path is None, Python is likely shutting down
Exception ignored in: <function RuntimeCache.__del__ at 0x7fe862091b20>
Traceback (most recent call last):
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 309, in __del__
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 204, in path
ImportError: sys.meta_path is None, Python is likely shutting down
Exception ignored in: <function RuntimeCache.__del__ at 0x7fe862091b20>
Traceback (most recent call last):
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 309, in __del__
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 204, in path
ImportError: sys.meta_path is None, Python is likely shutting down
Exception ignored in: <function RuntimeCache.__del__ at 0x7fe862091b20>
Traceback (most recent call last):
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 309, in __del__
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 204, in path
ImportError: sys.meta_path is None, Python is likely shutting down
Exception ignored in: <function RuntimeCache.__del__ at 0x7fe862091b20>
Traceback (most recent call last):
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 309, in __del__
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 204, in path
ImportError: sys.meta_path is None, Python is likely shutting down
Exception ignored in: <function RuntimeCache.__del__ at 0x7fe862091b20>
Traceback (most recent call last):
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 309, in __del__
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 204, in path
ImportError: sys.meta_path is None, Python is likely shutting down
Exception ignored in: <function RuntimeCache.__del__ at 0x7fe862091b20>
Traceback (most recent call last):
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 309, in __del__
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 204, in path
ImportError: sys.meta_path is None, Python is likely shutting down
Exception ignored in: <function RuntimeCache.__del__ at 0x7fe862091b20>
Traceback (most recent call last):
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 309, in __del__
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 204, in path
ImportError: sys.meta_path is None, Python is likely shutting down
Exception ignored in: <function RuntimeCache.__del__ at 0x7fe862091b20>
Traceback (most recent call last):
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 309, in __del__
  File "/home/TensorRT/py/torch_tensorrt/runtime/_runtime_cache.py", line 204, in path
ImportError: sys.meta_path is None, Python is likely shutting down
```

### PR DESCRIPTION
**Pull request by tp5uiuc**
## Summary
Original bug from #4359: engine-implicit `RuntimeCache` handles relied solely on `__del__` to autosave at process exit. When the engine survived to exit, `__del__` fired *during* shutdown and the torchbind / `filelock` lazy imports raised `ImportError: sys.meta_path is None`, surfacing as `Exception ignored in __del__` once per surviving handle.

### Fixes

- Move the autosave anchor to `atexit.register(partial(_autosave_at_exit,
  weakref.ref(self)))`. `atexit` fires before module teardown; `__del__`
  stays as the mid-program GC path. Both flip `autosave_on_del` off so
  whichever runs first wins. `__del__` unregisters its token to keep the
  atexit registry from accumulating dead entries.
- Module-level helper closing over a `weakref` keeps the registration
  non-owning, so handles can still die mid-program normally.
- `RuntimeCache.__getstate__` / `__setstate__` strip the unpicklable
  `weakref` from the pickle stream and re-register a fresh atexit hook
  on unpickle when `autosave_on_del` was on. `__del__` reads the token
  via `getattr(..., None)` so `copy.deepcopy` partial-init edge cases
  stay silent.

## Dependency on a sibling PR
This PR introduces an `_atexit_token` slot (`partial(..., weakref.ref(self))`) on `RuntimeCache`. CI tests that pickle compiled modules (`test_mutable_torchtrt_module::test_save`, `test_cross_runtime_serde::test_save_*`) walk through `_runtime_settings.runtime_cache` and would hit the weakref. The matching wrapper-side pickle exclusion lives in **#4368** and should land first to unblock CI here.

Fixes #4359.

## Type of change
- Bug fix (non-breaking)

## Test plan
`TestRuntimeCacheAutosave` in `tests/py/dynamo/runtime/test_000_runtime_cache.py`:
- [x] `test_del_swallows_shutdown_import_error_on_path` — `sys.unraisablehook` sees no leak
- [x] `test_atexit_hook_saves_via_weakref` — helper saves + flips flag
- [x] `test_atexit_hook_no_op_on_dead_weakref` — dead weakref no-ops cleanly
- [x] `test_atexit_token_unregistered_after_del` — `atexit.unregister` called with the registered token
- [x] `test_pickle_round_trip_strips_atexit_token` — standalone-pickle round-trip rewires a fresh hook (stubbed handle to isolate from the sibling PR's `_RuntimeCacheHandle` picklability work)
- [x] `pre-commit run` clean on touched files (mypy skipped for two pre-existing errors at `_TorchTensorRTModule.py:380` and `:508`, unrelated)

### CODE REVIEW

File: py/torch_tensorrt/runtime/_runtime_cache.py [hunk 1]
```diff
@@ -197,6 +214,17 @@ def __init__(
...
+        # down at interpreter exit -- ``__del__`` then hits ``ImportError``
+        # from the torchbind property and the lazy ``filelock`` import.
+        # ``atexit`` fires before that teardown. ``partial`` over a
+        # ``weakref`` keeps the registration non-owning, so mid-program GC
+        # still runs ``__del__`` normally.
+        if autosave_on_del:
+            self._atexit_token = atexit.register(
+                partial(_autosave_at_exit, weakref.ref(self))
+            )
+
```
Comment (tp5uiuc) [inline py/torch_tensorrt/runtime/_runtime_cache.py:225]: Can refactor this and the __setstate__ call into a single private function and reuse.

File: py/torch_tensorrt/runtime/_runtime_cache.py [hunk 2]
```diff
@@ -300,15 +328,67 @@ def save(self, path: Optional[str] = None) -> None:
...
+        ``Exception ignored in __del__``.
+        """
+        try:
+            if self.autosave_on_del and self.path:
+                self.autosave_on_del = False
+                self.save()
+        except Exception:
+            pass
+
+    def __getstate__(self) -> dict[str, Any]:
```
Comment (tp5uiuc) [inline py/torch_tensorrt/runtime/_runtime_cache.py:342]: Note to reviewers : these are added so that users calling pickle.save(RuntimeCache()) works as expected.

File: py/torch_tensorrt/runtime/_runtime_cache.py (hunk 1 above)
Comment (tp5uiuc) [reply py/torch_tensorrt/runtime/_runtime_cache.py:225]: [by Claude Code] Done in da18dc04e — hoisted to _register_atexit_autosave; __init__ and __setstate__ both call it; the idempotency check (`_atexit_token is None`) moved into the helper.

**Comment by tp5uiuc** [comment]
[by Claude Code] **CI status after rebase on #4367:**

- Wheel builds (Linux x86_64, SBSA, Windows, RTX, Python-only) all green now that #4367 fixed the `CUDAGraph::debug_dump` API mismatch.
- **`test_mutable_torchtrt_module::test_save` passes** — the regression target this PR was originally about. The `RuntimeCache.__getstate__` / `__setstate__` defense strips the `weakref` correctly without depending on #4368's wrapper exclusion.
- The 5 new `TestRuntimeCacheAutosave` tests are running and passing.

Remaining red is environmental noise hitting every concurrent PR (and main):
- `L0 dynamo converter` — `test_attention_aten::test_dynamic_sdpa_fp_mask_3` (torch-nightly SDPA API)
- `L0 torchscript` — `test_classes::test_get_layer_info` + worker crashes (pre-existing Windows torchscript)
- `Python-only` / `RTX Python-only dynamo runtime` — `test_004_weight_streaming` flakes

None of those touch files this PR modifies.

File: py/torch_tensorrt/runtime/_runtime_cache.py
Comment (cehongwang) [inline py/torch_tensorrt/runtime/_runtime_cache.py]: update the doc here

**Review by cehongwang — APPROVED** [review]
Minor comments

File: py/torch_tensorrt/runtime/_runtime_cache.py
Comment (tp5uiuc) [reply py/torch_tensorrt/runtime/_runtime_cache.py]: [by Claude Code] Done in d3d101252 — bullet 1 (Engine-implicit) now documents both the `__del__` and `atexit` hooks and the "whichever-fires-first flips `autosave_on_del` off" handshake. Compacted the `_handle` paragraph in the same commit.

Comment (tp5uiuc) [reply py/torch_tensorrt/runtime/_runtime_cache.py]: Thanks for the catch Adrian, I just added it.

**Comment by tp5uiuc** [comment]
[by Claude Code] **CI on d3d101252 (docs-only rebase):**

| Category | Result | Notes |
| --- | --- | --- |
| **Python-only dynamo runtime tests** (cu130, cu132) | ✅ pass | `test_mutable_torchtrt_module::test_save` still passing |
| **RTX Python-only dynamo runtime tests** (cu130, cu132) | ✅ pass | |
| L0 dynamo core / L0 core python tests | ✅ pass | |
| Build cu132 x86_64 + all Windows + all RTX + Jetpack wheels | ✅ pass | |
| Python + C++ Linting, Meta CLA | ✅ pass | |
| Build cu130 x86_64 + SBSA (cu130/cu132) wheels | 🔴 fail | `gitlab.com` overload cloning `eigen` submodule inside executorch — infra flake; cu132 x86_64 got through a few minutes later |
| L0 dynamo converter tests | 🔴 fail | pre-existing torch-nightly SDPA API mismatch (`test_dynamic_sdpa_fp_mask_3`), unrelated file |
| L0 torchscript tests | 🔴 fail | pre-existing Windows torchscript flakes, unrelated |

Regression target `test_mutable_torchtrt_module::test_save` and the 5 new `TestRuntimeCacheAutosave` tests all pass. Remaining red is a GitLab-side rate limit on `libeigen/eigen.git` plus known unrelated flakes — should clear on rerun.
